internal/identity/gitlab.go

```diff
@@ -0,0 +1,96 @@
+package identity // import "github.com/pomerium/pomerium/internal/identity"
+
+import (
+	"context"
+	"fmt"
+	"github.com/pomerium/pomerium/internal/httputil"
+	"github.com/pomerium/pomerium/internal/sessions"
+	"net/http"
+	"net/url"
+)
+import oidc "github.com/pomerium/go-oidc"
+import "golang.org/x/oauth2"
+
+// OIDCProvider provides a standard, OpenID Connect implementation
+// of an authorization identity provider.
+// https://openid.net/specs/openid-connect-core-1_0.html
+type GitLabProvider struct {
+	*Provider
+	Subject       string `json:"sub"`
+	Profile       string `json:"profile"`
+	Email         string `json:"email"`
+	EmailVerified bool   `json:"email_verified"`
+	// contains filtered or unexported fields
+}
+
+// NewGitLabProvider creates a new instance of a GitLab provider.
+func NewGitLabProvider(p *Provider) (*GitLabProvider, error) {
+	ctx := context.Background()
+	if p.ProviderURL == "" {
+		return nil, ErrMissingProviderURL
+	}
+	var err error
+	p.provider, err = oidc.NewProvider(ctx, p.ProviderURL)
+	if err != nil {
+		return nil, err
+	}
+	if len(p.Scopes) == 0 {
+		p.Scopes = []string{oidc.ScopeOpenID, "profile", "email", "offline_access"}
+	}
+	p.verifier = p.provider.Verifier(&oidc.Config{ClientID: p.ClientID})
+	p.oauth = &oauth2.Config{
+		ClientID:     p.ClientID,
+		ClientSecret: p.ClientSecret,
+		Endpoint:     p.provider.Endpoint(),
+		RedirectURL:  p.RedirectURL.String(),
+		Scopes:       p.Scopes,
+	}
+	return &GitLabProvider{Provider: p}, nil
+}
+
+// Revoke revokes the access token a given session state.
+// https://developer.okta.com/docs/api/resources/oidc#revoke
+//func (p *GitLabProvider) Revoke(ctx context.Context, token *oauth2.Token) error {
```

Comment:
It looks like gitlab supports token revocation so we should implement it if possible
https://docs.gitlab.com/ee/integration/oauth_provider.html